### ISSUE TITLE
`ChainNotify` erroneous behavior

### ISSUE THREAD

**Issue by alt-ivan** [ChainSafe/forest#4371]
## Describe the bug

There appears to be two bugs here:
1. `ChainNotify` yields one extra `Apply` change, same as `Current`
2. This only seems to happen for the first registered listener; the same change reported as `Current` is repeated once as Apply for the first registered listener, so multiple `ChainNotify` listeners get different results across for each change

See log for more details. The two lines per log entry are two `ChainNotify` WS readers 

## To reproduce

1. Run Lotus and Forest and connect them
4. Run one or more miners on Lotus
5. Attach two or more listeners via `ChainNotify` WS

## Log output

<details>
  <summary>Log Output</summary>
  
  ```Paste log output here
2024-05-28T09:36:25.865487Z  WARN lotus: Head change: current, Height: 838, CIDs bafy2bzacecc2u5qroitwqdavyj5rl6hejpnd5wqy5qhdrh57ct662uddrwtmw
2024-05-28T09:36:25.900517Z  WARN lotus: Head change: current, Height: 838, CIDs: bafy2bzacecc2u5qroitwqdavyj5rl6hejpnd5wqy5qhdrh57ct662uddrwtmw

2024-05-28T09:36:26.064284Z  WARN lotus: Head change: apply, Height: 839, CIDs: bafy2bzacedvrpxjktwd66ddhgvwuw35r7lrquihwwt4sjdrzy3y4j3wp2qso6
2024-05-28T09:36:26.065248Z  WARN lotus: Head change: apply, Height: 839, CIDs: bafy2bzacedvrpxjktwd66ddhgvwuw35r7lrquihwwt4sjdrzy3y4j3wp2qso6

2024-05-28T09:36:30.063523Z  WARN lotus: Head change: apply, Height: 840, CIDs: bafy2bzacea5hipinm2atf6payqyf4fljlemzq6srbcx3llt6lfubjzljvccfy
2024-05-28T09:36:30.064311Z  WARN lotus: Head change: apply, Height: 840, CIDs: bafy2bzacea5hipinm2atf6payqyf4fljlemzq6srbcx3llt6lfubjzljvccfy

...

2024-05-28T09:37:37.193983Z  WARN forest: Head change: current, Height: 856, CIDs: bafy2bzaceb5nsmc44vig5rny7gk7rxf5jwmmeacgn5lyqmeo77u37d3rzuycg
2024-05-28T09:37:37.194269Z  WARN forest: Head change: current, Height: 856, CIDs: bafy2bzaceb5nsmc44vig5rny7gk7rxf5jwmmeacgn5lyqmeo77u37d3rzuycg

2024-05-28T09:37:38.063049Z  WARN forest: Head change: apply, Height: 856, CIDs: bafy2bzaceb5nsmc44vig5rny7gk7rxf5jwmmeacgn5lyqmeo77u37d3rzuycg
2024-05-28T09:37:38.066762Z  WARN forest: Head change: apply, Height: 857, CIDs: bafy2bzaceaumuzo53m6vfojb4mdr7vsvwmco3ie3gvnurbbi6tjik2sfz2ly6

2024-05-28T09:37:42.074711Z  WARN forest: Head change: apply, Height: 857, CIDs: bafy2bzaceaumuzo53m6vfojb4mdr7vsvwmco3ie3gvnurbbi6tjik2sfz2ly6
2024-05-28T09:37:42.077869Z  WARN forest: Head change: apply, Height: 858, CIDs: bafy2bzacedmyovv23sn7nra5jwmltm5p5ot7pnkplnav7sat3ppb3tbzzhafw

2024-05-28T09:37:46.058826Z  WARN forest: Head change: apply, Height: 858, CIDs: bafy2bzacedmyovv23sn7nra5jwmltm5p5ot7pnkplnav7sat3ppb3tbzzhafw
2024-05-28T09:37:46.061476Z  WARN forest: Head change: apply, Height: 859, CIDs: bafy2bzacedw7ox4hxmuk7b6bitzk3toq7uoorj2aa45spwulhtenmffwtli6y


  ```
</details>

## Environment (please complete the following information):

- OS: Arch Linux via WSL2 :) (`5.15.90.1-microsoft-standard-WSL2`)
- Branch/commit https://github.com/ChainSafe/forest/pull/4357 merged `main` (`ca2d6a1b3d4d893273a93f176742a829d968bc67`)
- Hardware

**Comment by alt-ivan** [comment]
@elmattic 
Tested again using fc06b72da18851ee746a189ef1087b6dfe6f5a59, different behavior but still not in line with expected behavior:

```
2024-05-28T14:02:12.114294Z  WARN forest: Head change: current, Height: 36, CIDs: bafy2bzacebvolm4otol6fe7gkejgvhomgrevg2zlcv6fy352wzjyjujvc6wts
2024-05-28T14:02:12.115097Z  WARN forest: Head change: current, Height: 36, CIDs: bafy2bzacebvolm4otol6fe7gkejgvhomgrevg2zlcv6fy352wzjyjujvc6wts

2024-05-28T14:02:13.071703Z  WARN forest: Head change: apply, Height: 37, CIDs: bafy2bzacechpr7uw5ecso2hyrxdvlcphnhbx7of3qwtkjqqli7ii4sysdqtys

2024-05-28T14:02:17.068023Z  WARN forest: Head change: apply, Height: 37, CIDs: bafy2bzacechpr7uw5ecso2hyrxdvlcphnhbx7of3qwtkjqqli7ii4sysdqtys
2024-05-28T14:02:17.070717Z  WARN forest: Head change: apply, Height: 38, CIDs: bafy2bzaceb4a5tklninhw5tzt3fdoijxnku4hur72dy3fs75656dqp7fxfs5c

2024-05-28T14:02:21.060865Z  WARN forest: Head change: apply, Height: 38, CIDs: bafy2bzaceb4a5tklninhw5tzt3fdoijxnku4hur72dy3fs75656dqp7fxfs5c
2024-05-28T14:02:21.063914Z  WARN forest: Head change: apply, Height: 39, CIDs: bafy2bzacebbst45yvzdjp3ayfepzwkjfsehtpiypltw3vufxspwl2two4li5g

2024-05-28T14:02:25.060357Z  WARN forest: Head change: apply, Height: 39, CIDs: bafy2bzacebbst45yvzdjp3ayfepzwkjfsehtpiypltw3vufxspwl2two4li5g
2024-05-28T14:02:25.063661Z  WARN forest: Head change: apply, Height: 40, CIDs: bafy2bzaceburpw5lq5tkxy5erv5dt7ci6a6pm4uh2gpvbxnnh27uvhvulysw2

```

Now only one of the listeners gets the first `apply` change, and the two listeners still receive different changes and remain out of sync

**Comment by elmattic** [comment]
@alt-ivan Thanks again for the report. I'll try to reproduce it on a Linux machine. I don't observe this behavior on macOS and have what you expect.

**Comment by alt-ivan** [comment]
Confirming that the issue is fixed on #4357 so this issue may be closed at your convenience
